docs/RFCS/20210104_table_level_partitioning_and_regional_by_row_spec.md

```diff
@@ -0,0 +1,640 @@
+used as a unique identifier for rows in the table.
+
+> This requires that the values be both unique and not null. Adding a primary key
+will automatically create a unique B-tree index on the column or group of
+columns listed in the primary key.
+
+> There are also various ways in which the database system makes use of a
+primary key if one has been declared; for example, the primary key defines the
+default target column(s) for foreign keys referencing its table.
+
+In addition to these, there are also some more considerations regarding the index that
+accompanies a PK.
+
+### PK clustering semantics
+
+There’s one aspect of PKs on which databases differ: clustering semantics. In
+CRDB, currently, PKs are clustered indexes, meaning that scans over contiguous
+PK ranges are expected to be fast (and have access to all the columns in the
+table). Other databases differ in this regard:
+- CRDB - PK is (currently) clustered
+- Oracle - clustering optional, but only by PK
+- MySQL - PK is clustered
+- Postgres - PK not clustered, separate CLUSTER command
+- SQL Server - PK is clustered by default
+- DB2: PK not necessarily clustered, but “the first index” of a table is
+clustered? I don’t really understand it.
+
+Now, if you like to (continue to) assume that a PK is clustered, then you
+probably think that the PK should be `(crdb_region, id)` - because that’s the
+combination of columns that gives you fast range scans. If we “pretend” that
+product’s PK is just `(id)`, then, for example, a user might think that the
+following query will be fast when, in fact, it’ll be slow:
+```sql
+select * from product order by id limit 1
+```
+Or, if you want to continue advancing a cursor for a paginated query, you’d
+expect starting from an id and scanning forward in id order to be cheap.
+
+If we defined the table’s PK as simply `(id)`, then there’s arguably not even an
+index on `(id)` (in fact, the fact that Multi-region Functional Spec uses the
+`UNIQUE WITHOUT INDEX` clause for `(id)` uniqueness constraint spells out that
+there’s no index on it. In the context of regional by row tables, this “lack of
+an index” is muddled by the fact that, when data is partitioned across regions,
+users need to understand that indexes are not magic. Non-stale lookups by `(id)`
+and lookups by `(region, id)` need cross-region communication in the same
+circumstances (assuming a locality-optimized search is performed in the `(id)`
+case).
+
+
+## Detailed design
+We'll get to a solid definition of row-level regional tables in three steps:
+1. index partitioning improvements
+2. table-level partitioning
+3. build on the previous section to define row-level regional tables
+
+### Index partitioning improvements
+
+Consider this table definition:
+```sql
+CREATE TABLE t (
+region ENUM,
+id INT,
+x INT,
+PRIMARY KEY (region, id),
+INDEX(region, x) PARTITION BY LIST(region) (partition p1 values in (‘us-west’)...
```

Comment:
in theory, this isn't limited to just the region ENUM but could be any value (or even a set of values) right?